File: app/src/main/java/com/slapshotapps/dragonshockey/activities/roster/RosterFragment.java
package com.slapshotapps.dragonshockey.activities.roster;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.appcompat.widget.Toolbar;
import android.widget.TextView;
import butterknife.ButterKnife;
import com.google.firebase.database.DatabaseException;
import com.google.firebase.database.FirebaseDatabase;
import com.slapshotapps.dragonshockey.Config;
import com.slapshotapps.dragonshockey.R;
import com.slapshotapps.dragonshockey.ViewUtils.itemdecoration.StaticHeaderDecoration;
import com.slapshotapps.dragonshockey.activities.ActionBarListener;
import com.slapshotapps.dragonshockey.activities.HockeyFragment;
import com.slapshotapps.dragonshockey.activities.roster.adapters.RosterAdapter;
import com.slapshotapps.dragonshockey.models.Player;
import com.slapshotapps.dragonshockey.observables.RosterObserver;
import java.util.List;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action;
import rx.functions.Action1;
import rx.schedulers.Schedulers;
import timber.log.Timber;

/**
 * Created by willmetz on 7/25/16.
 */
public class RosterFragment extends HockeyFragment {

    private FirebaseDatabase firebaseDatabase;
    private Subscription rosterSubscription;
    private RecyclerView recyclerView;
    private TextView rosterUnavailable;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);

        firebaseDatabase = FirebaseDatabase.getInstance();

        try {
            firebaseDatabase.setPersistenceEnabled(true);
        }
        catch (DatabaseException exception) {
            Timber.e("Unable to set persistance for Firebase");
        }

        ActionBarListener listener = getActionBarListener();
        if(listener != null){
            listener.setTitle(getString(R.string.roster_title));
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.activity_roster, container, false);

        rosterUnavailable = view.findViewById(R.id.roster_unavailable);

        recyclerView = view.findViewById(R.id.roster_recycler_view);
        LinearLayoutManager layoutManager = new LinearLayoutManager(getContext(), RecyclerView.VERTICAL, false);
        recyclerView.setLayoutManager(layoutManager);

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();

        rosterSubscription = RosterObserver.GetRoster(firebaseDatabase)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .doOnError(new Action1<Throwable>() {
                @Override
                public void call(Throwable throwable) {
                    Timber.e("Unable to get the roster...");
                    rosterUnavailable.animate().alpha(1f);
                }
            })
            .subscribe(new Action1<List<Player>>() {
                @Override
                public void call(List<Player> players) {
                    rosterUnavailable.setAlpha(0);
                    RosterAdapter adapter =
                        new RosterAdapter(RosterFragment.this.getContext(), players, recyclerView);
                    recyclerView.setAdapter(adapter);

                    recyclerView.addItemDecoration(
                        new StaticHeaderDecoration(adapter, recyclerView));
                }
            });
    }

    @Override
    public void onPause() {
        super.onPause();

        if (rosterSubscription != null) {
            rosterSubscription.unsubscribe();
            rosterSubscription = null;
        }
    }
}

File: app/src/main/java/com/slapshotapps/dragonshockey/activities/roster/RosterFragment.kt
package com.slapshotapps.dragonshockey.activities.roster

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.slapshotapps.dragonshockey.R
import com.slapshotapps.dragonshockey.ViewUtils.itemdecoration.StaticHeaderDecoration
import com.slapshotapps.dragonshockey.activities.HockeyFragment
import com.slapshotapps.dragonshockey.activities.roster.adapters.RosterAdapter
import com.slapshotapps.dragonshockey.observables.RosterObserver
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import timber.log.Timber

/**
 * Created by willmetz on 7/25/16.
 */
class RosterFragment : HockeyFragment() {

    private var rosterSubscription: Subscription? = null
    private var recyclerView: RecyclerView? = null
    private var rosterUnavailable: TextView? = null

    override fun onAttach(context: Context) {
        super.onAttach(context)

        val listener = actionBarListener
        listener?.setTitle(getString(R.string.roster_title))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        analyticEventListener?.logContentSelectedEvent("Roster", "901")
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        val view = inflater.inflate(R.layout.activity_roster, container, false)

        rosterUnavailable = view.findViewById(R.id.roster_unavailable)

        recyclerView = view.findViewById(R.id.roster_recycler_view)
        val layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
        recyclerView!!.layoutManager = layoutManager

        return view
    }

    override fun onResume() {
        super.onResume()

        rosterSubscription = RosterObserver.GetRoster(firebaseDatabase)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnError {
                    Timber.e("Unable to get the roster...")
                    rosterUnavailable!!.animate().alpha(1f)
                }
                .subscribe { players ->
                    rosterUnavailable!!.alpha = 0f
                    val adapter = RosterAdapter(this@RosterFragment.context!!, players)
                    recyclerView!!.adapter = adapter

                    recyclerView!!.addItemDecoration(
                            StaticHeaderDecoration(adapter, recyclerView))
                }
    }

    override fun onPause() {
        super.onPause()

        if (rosterSubscription != null) {
            rosterSubscription!!.unsubscribe()
            rosterSubscription = null
        }
    }
}
